### PR TITLE
Improve error message for `impl Self`

### LINKED ISSUE #65784: Improve error message for `impl Self`
Suppose you have the following reduced code example:

```rust
impl Self {   
}
```

This gives you the error message:

```
error[E0391]: cycle detected when processing `<impl at src/lib.rs:1:1: 3:2>`
 --> src/lib.rs:1:6
  |
1 | impl Self {
  |      ^^^^
  |
  = note: ...which again requires processing `<impl at src/lib.rs:1:1: 3:2>`, completing the cycle
note: cycle used when collecting item types in top-level module
 --> src/lib.rs:1:1
  |
1 | impl Self {
  | ^^^^^^^^^
```

This is confusing because there isn't actually any obvious cycle (unless you know things about compilers). E0391 actually describes itself with an example like this:

```rust
trait FirstTrait : SecondTrait {

}

trait SecondTrait : FirstTrait {

}
```

This looks nothing like the original code example.

We should probably detect this case and provide a more useful message that indicates that `Self` is a special keyword that can be used within the impl block, but not as the type of the impl block itself.

### PR DESCRIPTION
**Pull request by TaKO8Ki**
closes #65784

### CODE REVIEW

File: compiler/rustc_resolve/src/late.rs [hunk 1]
```diff
@@ -1318,7 +1321,12 @@ impl<'a: 'ast, 'b, 'ast> LateResolutionVisitor<'a, 'b, 'ast> {
                             visit::walk_trait_ref(this, trait_ref);
                         }
                         // Resolve the self type.
+                        let prev = replace(
+                            &mut this.diagnostic_metadata.currently_processing_impl_self_ty,
+                            true,
+                        );
                         this.visit_ty(self_type);
```
Comment (cjgillot) [inline compiler/rustc_resolve/src/late.rs:1328]: What happens if we visit `self_type` outside of the call to `with_self_rib(res,`?

**Comment by compiler-errors** [comment]
You may also want to draw inspiration from #93971 which attempted to fix ICEs of similar variety.

File: compiler/rustc_resolve/src/late.rs (hunk 1 above)
Comment (TaKO8Ki) [reply compiler/rustc_resolve/src/late.rs:1328]: If we do, `visit_ty` just visits `self_ty` unless `currently_processing_impl_self_ty` is true. It never causes a `Self is only available in impls, traits, and type definitions` error.

In reference to https://github.com/rust-lang/rust/pull/93971#pullrequestreview-881042951, should I improve this error to report it right away?

File: src/test/ui/resolve/resolve-self-in-impl.rs [hunk 2]
```diff
@@ -12,9 +12,13 @@ impl Tr for S where S<Self>: Copy {} // OK
 impl Tr for S where Self::A: Copy {} // OK
 
 impl Tr for Self {} //~ ERROR cycle detected
+//~^ ERROR `Self` is only available in impls, traits, and type definitions
```
Comment (nagisa) [inline src/test/ui/resolve/resolve-self-in-impl.rs:15]: We should emit a single diagnostic for these, not multiple. I could see

> `Self` is only available in impls, traits, and type definitions

being a “help”, but I think it being a singular error makes more sense.

Comment (TaKO8Ki) [reply src/test/ui/resolve/resolve-self-in-impl.rs:15]: Ok. I'll fix it。

**Comment by JohnCSimon** [comment]
@TaKO8Ki  
Ping from triage: I'm closing this due to inactivity, Please reopen when you are ready to continue with this. 
Note: if you do please open the PR **BEFORE** you push to it, else you won't be able to reopen - this is a quirk of github. 
Thanks for your contribution.

@rustbot label: +S-inactive

**Pull request closed without merging**